### PR TITLE
Correct route precedence logic for k8s ingress

### LINKED ISSUE #35033: The behavior of istio ingress is different from the ingress documentation.
**Describe the feature request**

According to the [ingress documentation](https://kubernetes.io/docs/concepts/services-networking/ingress/#multiple-matches), it seems that routing is applied with longest matching path first. But the istio ingress(`kubernetes.io/ingress.class: istio`) seems to be affected by the order described in `paths` . For interoperability, I expect it to work as written in the Ingress documentation.

"In some cases, multiple paths within an Ingress will match a request. In those cases precedence will be given first to the longest matching path. If two paths are still equally matched, precedence will be given to paths with an exact path type over prefix path type."

https://github.com/goharbor/harbor-helm/issues/485#issuecomment-686222551

**Describe alternatives you've considered**

high priority to the longest matching path

**Affected product area (please put an X in all that apply)**

[X] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[X] Multi Cluster
[X] Virtual Machine
[X] Multi Control Plane

**Additional context**

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Sort generated vs routes to meet ingress route precedence requirements, please refer to https://kubernetes.io/docs/concepts/services-networking/ingress/#multiple-matches

> In some cases, multiple paths within an Ingress will match a request. In those cases **precedence will be given first to the longest matching path. If two paths are still equally matched, precedence will be given to paths with an exact path type over prefix path type.**

Fixes #35033

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Nice. Can we are a feature flag to retain the old behavior for backwards compatibility?

Also what happens for 2 ingress with same hostname?

**Comment by dddddai** [comment]
Thank you!
> Can we are a feature flag to retain the old behavior for backwards compatibility?

OK, so should the default value be false?

> Also what happens for 2 ingress with same hostname?

They will be merged to one vs whose routes are sorted, as the unit test shows:

ingress 1:
https://github.com/istio/istio/blob/af5e25e8615342de44f6816265574553f521848e/pilot/pkg/config/kube/ingressv1/conversion_test.go#L142-L149

ingress 2:
https://github.com/istio/istio/blob/af5e25e8615342de44f6816265574553f521848e/pilot/pkg/config/kube/ingressv1/conversion_test.go#L215-L222

result:
https://github.com/istio/istio/blob/af5e25e8615342de44f6816265574553f521848e/pilot/pkg/config/kube/ingressv1/conversion_test.go#L281-L283

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 1]
```diff
@@ -216,20 +216,22 @@ func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string,
 		if f {
 			vs := old.Spec.(*networking.VirtualService)
 			vs.Http = append(vs.Http, httpRoutes...)
-			sort.SliceStable(vs.Http, func(i, j int) bool {
```
Comment (dddddai) [inline pilot/pkg/config/kube/ingress/conversion.go:219]: And I'm a little confused about the old behavior, it seems to wanna sort the routes, but only when merging, so if there's only one ingress the generated vs routes will not be sorted, is that right?

**Comment by hzxuzhonghu** [comment]
nginx ingress-controller: 
https://github.com/nginxinc/kubernetes-ingress/tree/v2.1.0/examples/mergeable-ingress-types  

Also have CRD: https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/

Not all ingress controllers support ingress merge

**Comment by dddddai** [comment]
Thanks for the info
> Not all ingress controllers support ingress merge

Yes but we do merge ingresses
https://github.com/istio/istio/blob/994621ac3efbebbbadf1ed661a32d93e854d7508/pilot/pkg/config/kube/ingress/conversion.go#L134-L136

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 2]
```diff
@@ -216,20 +216,22 @@ func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string,
...
-				// TODO: default at the end
-				if r1Ex && !r2Ex {
-					return true
-				}
-				return false
-			})
 		} else {
 			ingressByHost[host] = &virtualServiceConfig
 		}
+
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/conversion.go:222]: for backwards compat we will need a flag for old behavior. Can combine it with https://github.com/istio/istio/pull/36855#pullrequestreview-855690740 maybe

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/ingress/conversion.go:219]: I think so. I think the idea was to ensure if we have 2 ingress with something like:
```
match0: /bar
match1: /
```

and

```
match0: /foo
match1: /
```

we put the `/` matches at the bottom so /foo and /bar can match

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 3]
```diff
@@ -216,20 +216,22 @@ func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string,
...
+		// sort routes to meet ingress route precedence requirements
+		// see https://kubernetes.io/docs/concepts/services-networking/ingress/#multiple-matches
+		vs := ingressByHost[host].Spec.(*networking.VirtualService)
+		sort.SliceStable(vs.Http, func(i, j int) bool {
+			r1Len, r1Ex := getMatchURILength(vs.Http[i].Match[0])
+			r2Len, r2Ex := getMatchURILength(vs.Http[j].Match[0])
+			// TODO: default at the end
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/conversion.go:229]: shouldn't default be at the end due to len=0?

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 4]
```diff
@@ -216,20 +216,22 @@ func ConvertIngressVirtualService(ingress v1beta1.Ingress, domainSuffix string,
...
+			if r1Len == r2Len {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingress/conversion.go:230]: > If two paths are still equally matched, precedence will be given to paths with an exact path type over prefix path type.

This is technically still ambiguous on which one wins... I guess here we are doing it based on creation time which seems reasonable

File: pilot/pkg/config/kube/ingressv1/conversion.go [hunk 5]
```diff
@@ -240,6 +242,22 @@ func ConvertIngressVirtualService(ingress knetworking.Ingress, domainSuffix stri
...
 }
 
+// getMatchURILength returns the length of matching path, and whether the match type is EXACT
+func getMatchURILength(match *networking.HTTPMatchRequest) (length int, exact bool) {
+	uri := match.GetUri()
+	switch uri.GetMatchType().(type) {
+	case *networking.StringMatch_Exact:
+		return strings.Count(uri.GetExact(), "/"), true
+	case *networking.StringMatch_Prefix:
+		return strings.Count(uri.GetPrefix(), "/"), false
```
Comment (howardjohn) [inline pilot/pkg/config/kube/ingressv1/conversion.go:252]: Is it the number of `/` or the raw `len(uri.GetPrefix())` we want? Can we look at some others like nginx and countor and see what they do?

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 4 above)
Comment (dddddai) [reply pilot/pkg/config/kube/ingress/conversion.go:230]: We already sorted the ingresses based on creation time:
https://github.com/istio/istio/blob/994621ac3efbebbbadf1ed661a32d93e854d7508/pilot/pkg/config/kube/ingress/controller.go#L348

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 3 above)
Comment (dddddai) [reply pilot/pkg/config/kube/ingress/conversion.go:229]: The comment is derived from [here](https://github.com/istio/istio/pull/36847/files#diff-b01367d947f649ce053caa5a9ffda6141cf242cfda73cfd62687fe185603e689L224)

I guess "default" means `pathType: ImplementationSpecific`?

File: pilot/pkg/config/kube/ingress/conversion.go (hunk 2 above)
Comment (dddddai) [reply pilot/pkg/config/kube/ingress/conversion.go:234]: Done

File: pilot/pkg/config/kube/ingressv1/conversion.go (hunk 5 above)
Comment (dddddai) [reply pilot/pkg/config/kube/ingressv1/conversion.go]: Oh yes, I've checked [ingress-nginx doc](https://kubernetes.github.io/ingress-nginx/user-guide/ingress-path-matching/#example_1), should use the raw length.
Fixed

**Review by howardjohn — APPROVED** [review]

**Pull request merged**